controllers/keyword_test.go

```diff
@@ -0,0 +1,131 @@
+package controllers
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"testing"
+
+	"github.com/bxcodec/faker/v3"
+	"github.com/gin-gonic/gin"
+	"github.com/gutakk/go-google-scraper/db"
+	"github.com/gutakk/go-google-scraper/models"
+	testConfig "github.com/gutakk/go-google-scraper/tests/config"
+	testDB "github.com/gutakk/go-google-scraper/tests/db"
+	testFile "github.com/gutakk/go-google-scraper/tests/file"
+	testHttp "github.com/gutakk/go-google-scraper/tests/http"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/crypto/bcrypt"
+	"gopkg.in/go-playground/assert.v1"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+type KeywordDbTestSuite struct {
+	suite.Suite
+	engine *gin.Engine
+	cookie string
+}
+
+func (s *KeywordDbTestSuite) SetupTest() {
+	database, _ := gorm.Open(postgres.Open(testDB.ConstructTestDsn()), &gorm.Config{})
+	db.GetDB = func() *gorm.DB {
+		return database
+	}
+
+	_ = db.GetDB().AutoMigrate(&models.User{}, &models.Keyword{})
+
+	s.engine = testConfig.GetRouter(true)
+	new(LoginController).applyRoutes(EnsureGuestUserGroup(s.engine))
+	new(KeywordController).applyRoutes(EnsureAuthenticatedUserGroup(s.engine))
+
+	email := faker.Email()
+	password := faker.Password()
+
+	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	user := models.User{Email: email, Password: string(hashedPassword)}
+	db.GetDB().Create(&user)
+
+	formData := url.Values{}
+	formData.Set("email", email)
+	formData.Set("password", password)
+
+	headers := http.Header{}
+	headers.Set("Content-Type", "application/x-www-form-urlencoded")
+	response := testHttp.PerformRequest(s.engine, "POST", "/login", headers, formData)
```

Comment:
Sorry for not noticing this before, but having to manually log in for a unit tests seem awkward and not standard 🤔 Wouldn't saving a cookie as a fixture and set it manually would not work? Let me know if this is because of a limitation of Gin.

Reply:
Good catch. Finally I found the way to save a cookie as a fixture in 4f3d586812302dec2af69154f873a02ea3475a22